packages/@react-aria/collections/src/Document.ts

```diff
@@ -470,6 +473,12 @@ export class Document<T, C extends BaseCollection<T> = BaseCollection<T>> extend
     }
 
     // Next, update dirty collection nodes.
+    // TODO: when updateCollection is called here, shouldn't we be calling this.updateChildIndicies as well? Theoretically it should only update
+    // nodes from _minInvalidChildIndex onwards so the increase in dirtyNodes should be minimal.
+    // Is element.updateNode supposed to handle that (it currently assumes the index stored on the node is correct already).
+    // At the moment, without this call to updateChildIndicies, filtering an async combobox doesn't actually update the index values of the
+    // updated collection...
+    this.updateChildIndices();
```

Comment:
This change along with the above were mainly to fix an issue I noticed where the collection indexes weren't updating when filtering the combobox. Above, we properly update the `childIndices` of the Document's dirty nodes (aka the children's children) but not for the main children. Not sure if the call to `element.updateNode` below is supposed to handle this but this felt the most correct

Reply:
The todo is part of the comment above, basically just noting that I'm not 100% sure about the change. From what I observed the document never marks itself as dirty when performing `insertBefore` operations, only when calling `appendChild` or if `setFirstChild` gets hit in `insertBefore`. Instead of this change, I could add `  this.ownerDocument.markDirty(this);` to BaseNode's `insertBefore` and `removeChild` but it felt safer to isolate it here to `updateCollection` as a "end step" of sorts